vowpalwabbit/core/src/reductions/csoaa_ldf.cc

```diff
@@ -499,10 +499,61 @@ const VW::example* get_example_with_labelled_class(uint32_t class_to_find, const
   return nullptr;
 }
 
-void output_probabilities_example(VW::workspace& all, const VW::multi_ex& ec_seq)
+void end_pass(ldf&) {}
+
+void update_stats_csoaa_ldf_rank(const VW::workspace& /* all */, VW::shared_data& sd, const ldf& /* data */,
+    const VW::multi_ex& ec_seq, VW::io::logger& logger)
 {
+  if (test_ldf_sequence(ec_seq, logger)) { sd.weighted_unlabeled_examples += ec_seq[0]->weight; }
+  else { sd.weighted_labeled_examples += ec_seq[0]->weight; }
+  sd.example_number++;
+  const auto& head_ec = *ec_seq[0];
+
   size_t num_features = cs_count_features(ec_seq);
-  all.sd->total_features += num_features;
+  sd.total_features += num_features;
+
+  const VW::v_array<VW::action_score>& preds = head_ec.pred.a_s;
+
+  const auto* predicted_ec = get_example_with_labelled_class(preds[0].action, ec_seq);
+  if (predicted_ec != nullptr)
+  {
+    const auto loss = predicted_ec->l.cs.costs[0].x;
+    assert(loss >= 0);
+    sd.sum_loss += loss;
+    sd.sum_loss_since_last_dump += loss;
+  }
+}
+
+void output_example_prediction_csoaa_ldf_rank(
+    VW::workspace& all, const ldf& /* data */, const VW::multi_ex& ec_seq, VW::io::logger& logger)
+{
+  const auto& head_ec = *ec_seq[0];
+  for (auto& sink : all.final_prediction_sink)
+  {
+    VW::details::print_action_score(sink.get(), head_ec.pred.a_s, head_ec.tag, logger);
+  }
+  if (all.raw_prediction != nullptr) { csoaa_ldf_print_raw(all, all.raw_prediction.get(), ec_seq, logger); }
+  VW::details::global_print_newline(all.final_prediction_sink, logger);
+}
+
+void print_update_csoaa_ldf_rank(VW::workspace& all, VW::shared_data& /* sd */, const ldf& /* data */,
+    const VW::multi_ex& ec_seq, VW::io::logger& /* unused */)
+{
+  const auto& head_ec = *ec_seq[0];
+  const VW::v_array<VW::action_score>& preds = head_ec.pred.a_s;
+  const auto* predicted_ec = get_example_with_labelled_class(preds[0].action, ec_seq);
+  VW::details::print_cs_update_action_scores(all, predicted_ec == nullptr, cs_count_features(ec_seq), preds);
+}
+
+void update_stats_csoaa_ldf_prob(const VW::workspace& all, VW::shared_data& sd, const ldf& /* data */,
+    const VW::multi_ex& ec_seq, VW::io::logger& logger)
+{
+  if (test_ldf_sequence(ec_seq, logger)) { sd.weighted_unlabeled_examples += ec_seq[0]->weight; }
+  else { sd.weighted_labeled_examples += ec_seq[0]->weight; }
+  sd.example_number++;
+
+  size_t num_features = cs_count_features(ec_seq);
+  sd.total_features += num_features;
```

Comment:
the fact that the shared_data object can be shared across multiple processes without any kind of synchronization terrifies me